File: advent-of-code/src/main/kotlin/com/willmolloy/adventofcode/_2025/Day2.kt
package com.willmolloy.adventofcode._2025

import com.willmolloy.adventofcode.common.Day
import com.willmolloy.adventofcode.common.Input

/** https://adventofcode.com/2025/day/2 */
object Day2 : Day(2025, 2) {

  override fun part1(input: Input): Any {
    val split = input.string().split(",")

    var res  = 0L

    for (s in split) {
      val r1 = s.split("-")[0].toLong()
      val r2 = s.split("-")[1].toLong()

      for (i in r1..r2) {
        val iString = i.toString()
        val first = i.toString().substring(0, iString.length / 2)
        val second = i.toString().substring(iString.length / 2)
        if (first == second){
          res += i
        }
      }
    }

    return res
  }

  override fun part2(input: Input): Any {
    val split = input.string().split(",")

    var res  = 0L

    for (s in split) {
      val r1 = s.split("-")[0].toLong()
      val r2 = s.split("-")[1].toLong()

      for (i in r1..r2) {
        val iString = i.toString()

        for (j in 1 until iString.length) {
          val chunked = iString.chunked(j)
          if (chunked.all { s -> s == chunked[0] }){
            res += i
            break;
          }
        }
      }
    }

    return res
  }
}

File: advent-of-code/src/main/scala/com/willmolloy/adventofcode/_2024/Day2.scala
package com.willmolloy.adventofcode._2024

import com.willmolloy.adventofcode.common.{Day, Input}
import com.willmolloy.adventofcode.scala.Adapters.ListAdapter
import com.willmolloy.adventofcode.scala.Adapters.ArrayAdapter

import scala.annotation.tailrec

/**
 * https://adventofcode.com/2024/day/2
 */
object Day2 extends Day(2024, 2) {

  override def part1(input: Input): Int = {
    input.readLinesOfNums().count(isSafe)
  }

  override def part2(input: Input): Int = {
    // try all combinations
    input
      .readLinesOfNums()
      .count(report =>
        report.indices.exists(i => {
          val newReport = report.take(i) ++ report.drop(i + 1)
          isSafe(newReport)
        })
      )
  }

  private def isSafe(report: Seq[Long]): Boolean = {
    @tailrec
    def isSafe(list: Seq[Long], increasing: Boolean): Boolean = {
      list match
        case Seq() | Seq(_) => true
        case a +: b +: tail => {
          if (increasing && a >= b || !increasing && a <= b) {
            return false
          }

          val diff = math.abs(a - b)
          if (diff < 1 || diff > 3) {
            return false
          }

          isSafe(b +: tail, increasing)
        }
    }

    if (report.length < 2) true
    else isSafe(report, report(0) < report(1))
  }
}
